src/main/java/edu/harvard/iq/dataverse/api/Datasets.java

```diff
@@ -2502,6 +2506,76 @@ public Response addFileToDataset(@PathParam("id") String idSupplied,
     } // end: addFileToDataset
 
 
+    /**
+     * Clean storage of a Dataset
+     *
+     * @param idSupplied
+     * @return
+     */
+    @GET
+    @Path("{id}/cleanStorage")
+    public Response cleanStorage(@PathParam("id") String idSupplied, @QueryParam("dryrun") Boolean dryrun) {
+        // get user and dataset
+        User authUser;
+        try {
+            authUser = findUserOrDie();
+        } catch (WrappedResponse ex) {
+            return error(Response.Status.FORBIDDEN,
+                    BundleUtil.getStringFromBundle("file.addreplace.error.auth")
+            );
+        }
+
+        Dataset dataset;
+        try {
+            dataset = findDatasetOrDie(idSupplied);
+        } catch (WrappedResponse wr) {
+            return wr.getResponse();
+        }
+        
+        // check permissions
+        if (!permissionSvc.permissionsFor(createDataverseRequest(authUser), dataset).contains(Permission.EditDataset)) {
+            return error(Response.Status.INTERNAL_SERVER_ERROR, "Access denied!");
+        }
+
+        boolean doDryRun = dryrun != null && dryrun.booleanValue();
+
+        // check if no legacy files are present
+        Set<String> datasetFilenames = getDatasetFilenames(dataset);
+        if (datasetFilenames.stream().anyMatch(x -> !dataFilePattern.matcher(x).matches())) {
+            return error(Response.Status.INTERNAL_SERVER_ERROR, "Dataset contains files not matching the nameing pattern!");
```

Comment:
> first, I ensure that all the files in the dataset follow the regex pattern (no legacy files present), otherwise an error is thrown ... 

But, do we really need to do this? I mean, just because it's a potentially ancient dataset (and/or a dataset that may have been modified in some custom way) doesn't mean that we are not going to find any recent junk files that our regex is going to match. So, it is possible that we may still be able to recover some wasted storage space - ? 

Or do you feel that it's just safer, to communicate to the admin that "this is a weird/non-standard dataset, we don't want to touch it, you're on your own maintaining it?"

Reply:
How I see it, is that it is no problem to let the cleanup do its thing on the legacy datasets. It should never remove something that may not be removed: the regex does not allow it. The only reason to not do that is that it could give a false sense of data being clean. A warning i.s.o. error sounds good. I will change the code that it gives a warning about the legacy files → these will not be cleaned, only the files following the naming conventions will be cleaned.